### PR TITLE
Integrate llvm-project @864902e9b4d8bc6d3f0852d5c475e3dc97dd8335

### PR DESCRIPTION
**Pull request by Max191**
bump llvm-project to https://github.com/llvm/llvm-project/commit/864902e9b4d8bc6d3f0852d5c475e3dc97dd8335

Main changes in IREE (causing test updates):
- New forall loop canonicalizations are added which fold induction variables. This can canonicalilze away some index computation, and fold some extract_slice ops, so a few pipeline tests are changed.
- A new option for `hoistRedundantVectorTransfers` (`verifyNonZeroTrip`) is set, which prevents hoisting from possibly zero trip loops. This option is flipped in this integrate because it fixes a bug exposed by forall loop canonicalizations.

### CODE REVIEW

**Review by kuhar — COMMENTED** [review]
How does the performance look like after this integrate?

File: compiler/src/iree/compiler/Codegen/Common/OptimizeVectorTransferPass.cpp [hunk 1]
```diff
@@ -91,7 +91,8 @@ struct OptimizeVectorTransferPass final
       // Workaround, run loop invariant code motion before hoist redundant
       // vector transfer to workaround a bug upstream.
       loopInvariantCodeMotion(funcOp);
-      linalg::hoistRedundantVectorTransfers(cast<func::FuncOp>(funcOp));
+      linalg::hoistRedundantVectorTransfers(cast<func::FuncOp>(funcOp),
+                                            /*verifyNonZeroTrip=*/true);
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/Common/OptimizeVectorTransferPass.cpp:95]: We can make this false for the integrate which keeps the same behavior in the integrate commit. Then you can send another patch to enable this.

**Comment by Max191** [comment]
> How does the performance look like after this integrate?

Based on the CI, all the SDXL numbers are down from their golden numbers (e2e ~370ms->340ms). I'm not sure if these golden numbers are up to date, though.

File: compiler/src/iree/compiler/Codegen/Common/OptimizeVectorTransferPass.cpp (hunk 1 above)
Comment (Max191) [reply compiler/src/iree/compiler/Codegen/Common/OptimizeVectorTransferPass.cpp:95]: The issue is that this test fails after the loop canonicalization patch that comes in from this integrate:
https://github.com/iree-org/iree/blob/868768dd328678f12e01359a4e8fb5bb839c00a1/compiler/src/iree/compiler/Codegen/LLVMGPU/test/ROCDL/pipeline_tile_and_fuse.mlir#L1070-L1074

I added the `verifyNonZeroTrip` option upstream to handle this case along with the loop canonicalizations.

Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/Common/OptimizeVectorTransferPass.cpp:95]: Oh I see, there are other changes already break the tests and the change here fixes the issue.

**Review by hanhanW — APPROVED** [review]
